lib/src/folding/parsers/abstract.dart

```diff
@@ -0,0 +1,244 @@
+  /// - Ends with import or possibleImport line.
+  /// - Has at least one import line.
+  /// - Only can have import, possibleImport, blank, singleLineComment lines.
+  ///
+  /// [lineIndex] is the current line that does not have import statements.
+  /// So the latest possible import line is [lineIndex] - 1.
+  @protected
+  void endImportSequenceIfAny(int lineIndex) {
+    // We can have many possibleImport lines that add to the block range,
+    // but at least one true import line is required for the block.
+    bool importFound = false;
+    int? first; // Min line index of the block.
+    int? last; // Max line index of the block.
+
+    // Called if the block upper bound is found. Adds if it is a valid block.
+    void addIfFound() {
+      if (first == null || !importFound || last == first) {
+        return;
+      }
+
+      blocks.add(
+        FoldableBlock(
+          startLine: first,
+          endLine: last!,
+          type: FoldableBlockType.imports,
+        ),
+      );
+    }
+
+    int i = lineIndex;
+
+    for (final line in _linesWithSemantics.reversed) {
+      if (line.index != --i) {
+        // A gap in submitted lines. This happens with uninterpreted blocks.
+        // They should break the block.
+        addIfFound();
+        return;
+      }
+
+      if (line.semantics == LineSemantics.import) {
+        importFound = true;
+      }
+
+      switch (line.semantics) {
+        case LineSemantics.import:
+        case LineSemantics.possibleImport:
+          first = line.index;
+          last ??= line.index;
+          break;
+
+        case LineSemantics.singleLineCommentAndImportTerminator:
+          addIfFound();
+          return;
+
+        case LineSemantics.blank:
+        case LineSemantics.singleLineComment:
+        case LineSemantics.singleLineCommentTerminator:
+          break; // no op
+      }
+    }
+
+    addIfFound();
+  }
+
+  /// Terminates a single-line comment sequence if one was started above.
+  ///
+  /// A single-line comment block is a block that:
+  /// - Starts and ends with lines with [LineSemantics.singleLineComment].
+  /// - Can only contain [LineSemantics.singleLineComment]
+  ///   and [LineSemantics.blank] inside.
+  ///
+  /// [lineIndex] is the current line that is not
+  /// [LineSemantics.singleLineComment].
+  /// So the latest possible block line is [lineIndex] - 1.
+  @protected
+  void endCommentSequenceIfAny(int lineIndex) {
+    int? first; // Min line index of the block.
+    int? last; // Max line index of the block.
+
+    // Called if the block upper bound is found. Adds if it is a valid block.
+    void addIfFound() {
+      if (first == null || last == first) {
+        return;
+      }
+
+      blocks.add(
+        FoldableBlock(
+          startLine: first,
+          endLine: last!,
+          type: FoldableBlockType.singleLineComment,
+        ),
+      );
+    }
+
+    int i = lineIndex;
+
+    for (final line in _linesWithSemantics.reversed) {
```

Comment:
endCommentSequenceIfAny, endImportSequenceIfAny and other similar functions seem too big. Can they be broken down into smaller ones?

Reply:
I can't think of a way that would not also make it slower.

This will be called on every character change for every line in the code, so it may be critical, although the performance is not tested.

This code complexity could be tolerated for code covered well.